sklearn/utils/stats.py

```diff
@@ -32,41 +37,63 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     percentile : int if `array` 1D, ndarray if `array` 2D
         Weighted percentile.
     """
+    xp, is_array_api_compliant, device = get_namespace_and_device(array)
+    sample_weight = xp.asarray(
+        sample_weight,
+        dtype=_find_matching_floating_dtype(sample_weight, xp=xp),
+        device=device,
+    )
     n_dim = array.ndim
     if n_dim == 0:
         return array[()]
     if array.ndim == 1:
-        array = array.reshape((-1, 1))
+        array = xp.reshape(array, (-1, 1))
     # When sample_weight 1D, repeat for each array.shape[1]
     if array.shape != sample_weight.shape and array.shape[0] == sample_weight.shape[0]:
-        sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
-    sorted_idx = np.argsort(array, axis=0)
-    sorted_weights = np.take_along_axis(sample_weight, sorted_idx, axis=0)
+        sample_weight = xp.tile(sample_weight, (array.shape[1], 1)).T
+    sorted_idx = xp.argsort(array, axis=0)
+    sorted_weights = _take_along_axis(sample_weight, sorted_idx, xp)
 
     # Find index of median prediction for each sample
     weight_cdf = stable_cumsum(sorted_weights, axis=0)
     adjusted_percentile = percentile / 100 * weight_cdf[-1]
+    weight_cdf = xp.asarray(
+        weight_cdf, dtype=_find_matching_floating_dtype(weight_cdf, xp=xp)
+    )
+    adjusted_percentile = xp.asarray(
+        adjusted_percentile,
+        dtype=_find_matching_floating_dtype(adjusted_percentile, xp=xp),
+    )
 
     # For percentile=0, ignore leading observations with sample_weight=0. GH20528
     mask = adjusted_percentile == 0
-    adjusted_percentile[mask] = np.nextafter(
-        adjusted_percentile[mask], adjusted_percentile[mask] + 1
+    adjusted_percentile[mask] = _nextafter(
+        adjusted_percentile[mask], adjusted_percentile[mask] + 1, xp=xp
     )
 
-    percentile_idx = np.array(
+    percentile_idx = xp.asarray(
         [
-            np.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
+            xp.searchsorted(weight_cdf[:, i], adjusted_percentile[i])
             for i in range(weight_cdf.shape[1])
         ]
     )
-    percentile_idx = np.array(percentile_idx)
+
     # In rare cases, percentile_idx equals to sorted_idx.shape[0]
     max_idx = sorted_idx.shape[0] - 1
-    percentile_idx = np.apply_along_axis(
-        lambda x: np.clip(x, 0, max_idx), axis=0, arr=percentile_idx
+    percentile_idx = _apply_along_axis(
+        lambda x: xp.clip(x, 0, max_idx), axis=0, arr=percentile_idx, xp=xp
     )
```

Comment:
I am wondering, is `apply_along_axis` really needed here?

How can this not yield the same result as directly calling the following?

```python
xp.clip(percentile_idx, 0, max_idx)
```

Reply:
This would allow us not implement a complex and slow generic implementation for `_apply_along_axis`.

Reply:
I agree, because we are just applying the `clip` function along the 0 axis. Going to try this and see if it makes any difference